taichi/transforms/offload.cpp

```diff
@@ -426,18 +441,90 @@ class FixCrossOffloadReferences : public BasicStmtVisitor {
     local_to_global_vector_type[stmt] = ret_type;
     auto ptr = replacement.push_back<GlobalTemporaryStmt>(
         local_to_global_offset[stmt], ret_type);
-    LaneAttribute<TypedConstant> zeros(std::vector<TypedConstant>(
-        stmt->width(), TypedConstant(stmt->ret_type)));
-    auto const_zeros = replacement.push_back<ConstStmt>(zeros);
-    replacement.push_back<GlobalStoreStmt>(ptr, const_zeros);
+    if (auto tensor_type = stmt->ret_type->cast<TensorType>()) {
+      LaneAttribute<TypedConstant> zero(std::vector<TypedConstant>(
+          1, TypedConstant(tensor_type->get_element_type())));
+      auto const_zero_stmt = replacement.push_back<ConstStmt>(zero);
+      for (int i = 0; i < tensor_type->get_num_elements(); ++i) {
+        LaneAttribute<TypedConstant> offset(std::vector<TypedConstant>(
+            1, TypedConstant(i *
+                             data_type_size(tensor_type->get_element_type()))));
+        auto const_offset_stmt = replacement.push_back<ConstStmt>(offset);
+        auto ptr_offset_stmt =
+            replacement.push_back<PtrOffsetStmt>(ptr, const_offset_stmt);
+        replacement.push_back<GlobalStoreStmt>(ptr_offset_stmt,
+                                               const_zero_stmt);
+      }
+    } else {
+      LaneAttribute<TypedConstant> zeros(std::vector<TypedConstant>(
+          stmt->width(), TypedConstant(stmt->ret_type)));
+      auto const_zeros = replacement.push_back<ConstStmt>(zeros);
+      replacement.push_back<GlobalStoreStmt>(ptr, const_zeros);
+    }
 
     stmt->parent->replace_with(stmt, std::move(replacement), false);
     throw IRModified();
   }
 
+  void visit(PtrOffsetStmt *stmt) override {
+    auto alloca = stmt->origin;
+    auto offset = stmt->offset;
+
+    if (stmt_to_offloaded[offset] != stmt_to_offloaded[stmt]) {
+      VecStatement replacement;
+      // TODO: offset may not be ConstStmt
+      auto copy_stmt = offset->as<ConstStmt>()->copy();
+      auto copy = replacement.push_back(std::move(copy_stmt));
+      stmt_to_offloaded[copy] = stmt_to_offloaded[stmt];
+      auto ptr_offset = replacement.push_back<PtrOffsetStmt>(alloca, copy);
+      stmt->parent->replace_with(stmt, std::move(replacement));
+      throw IRModified();
+    }
+
+    if (!(alloca->is<AllocaStmt>() &&
+          alloca->cast<AllocaStmt>()->ret_type->is<TensorType>()))
+      return;
+    if (local_to_global_offset.find(alloca) == local_to_global_offset.end())
+      return;
+
+    VecStatement replacement;
+    auto ret_type = alloca->cast<AllocaStmt>()->ret_type;
+
+    auto ptr = replacement.push_back<GlobalTemporaryStmt>(
+        local_to_global_offset[alloca], ret_type);
+    replacement.push_back<PtrOffsetStmt>(ptr, stmt->offset);
+
+    stmt->parent->replace_with(stmt, std::move(replacement));
+    throw IRModified();
+  }
+
   // Replace local LD/ST with global LD/ST
   void visit(LocalLoadStmt *stmt) override {
     TI_ASSERT(stmt->width() == 1);
+    // TensorType Alloca
```

Comment:
Could you create a function for these three Stmts?

Reply:
This has been done in another PR for simplifying offload pass and supporting the first issue.